analysis/pfe_methods/range_request_pfe_method.py

```diff
@@ -0,0 +1,206 @@
+"""Implementation of PFE that downloads individual glyphs selectively.
+
+This PFE method downloads everything but the outlines,
+runs shaping on the client to determine which glyphs are necessary,
+and then uses range requests to download the specific glyphs which are necessary.
+"""
+
+import io
+import objc
+import Foundation
+import CoreText
+import AppKit
+
+from analysis import network_models
+from analysis import request_graph
+from collections import defaultdict
+from fontTools import ttLib
+
+GLYPH_SIZE_CACHE = dict()
+
+# Experimentally gathered by gzipping all the fonts in the Google Fonts corpus which support Chinese
+# FIXME: Gather the actual glyf bytes so we can gzip the real data instead of using this average
+AVERAGE_GZIP_COMPRESSION_RATIO = 0.462730945370997
+
+def name():
+  return "RangeRequest"
+
+def initialize():
+  AppKit.NSFont.initialize()
+
+# Possible ideas for network_startup_cost_in_bytes:
+# network_model.rtt * network_model.bandwidth_down
+# network_models.ESTIMATED_HTTP_REQUEST_HEADER_SIZE + network_models.ESTIMATED_HTTP_RESPONSE_HEADER_SIZE
+# Whatever it has to be such that (network_model.rtt * network_model.bandwidth_down) / len(requests) == network_startup_cost_in_bytes
+def start_session(font_loader, network_startup_cost_in_bytes=network_models.ESTIMATED_HTTP_REQUEST_HEADER_SIZE + network_models.ESTIMATED_HTTP_RESPONSE_HEADER_SIZE):
+  return RangeRequestPfeSession(font_loader, network_startup_cost_in_bytes)
+
+def codepoints_to_glyphs(font_bytes, codepoints):
+  font = ttLib.TTFont(io.BytesIO(font_bytes))
+  cmap = font["cmap"].getBestCmap()
+  return set([font.getGlyphID(cmap[codepoint]) for codepoint in codepoints if codepoint in cmap])
+
+class RangeRequestPfeSession:
+
+  def __init__(self, font_loader, network_startup_cost_in_bytes):
+    initialize()
+    # FIXME: network_startup_cost_in_bytes should be gathered from the actual network models.
+    self.font_loader = font_loader
+    self.network_startup_cost_in_bytes = network_startup_cost_in_bytes
+    self.request_graphs = []
+    self.loaded_glyphs = defaultdict(set)
+    # FIXME: Build this framework as part of the project.
+    bundle = objc.loadBundle("Optimizer", globals(), bundle_path="/Users/mmaxfield/Build/Products/Debug/Optimizer.framework")
+    self.GlyphSizeComputer = bundle.classNamed_("ExposedGlyphSizeComputer")
+
+  def compute_glyph_sizes(self, font_id):
+    self.font_loader.load_font(font_id) # Guard against invalid fonts
+    data = Foundation.NSData.dataWithContentsOfFile_(self.font_loader.path_for_font(font_id))
+    fontDescriptor = CoreText.CTFontManagerCreateFontDescriptorFromData(data)
+    font = CoreText.CTFontCreateWithFontDescriptor(fontDescriptor, 0, None)
+    computer = self.GlyphSizeComputer.alloc().initWithFont_(font)
+    return computer.computeGlyphSizes()
+
+  def get_base_font_size(self, font_sizes):
+    return font_sizes.fontSize() - sum(font_sizes.glyphSizes())
+
+  class GlyphRange:
```

Comment:
Because this looks like a coding style choice and has sat here unanswered for a while, I think it best to merge  as-is because it is more valuable to get some analysis results frm this PFE method than to worry about longer term code maintainability

Reply:
> Named tuples are immutable

https://stackoverflow.com/a/19257700